File: go/0041_first_missing.positive.go
package main

func firstMissingPositive(nums []int) int {
	nlen := len(nums)
	for i := 0; i < nlen; i++ {
		if nums[i] >= 1 && nums[i] <= nlen && nums[i] != nums[nums[i]-1] {
			nums[i], nums[nums[i]-1] = nums[nums[i]-1], nums[i]
			i-- // keep the same index
		}
	}

	for i := 0; i < nlen; i++ {
		if nums[i] != i+1 {
			return i + 1
		}
	}
	return nlen + 1
}

File: python/0041_first_missing_positive.py
class Solution:
    def firstMissingPositive(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """

        nlen = len(nums)
        if nlen == 0:
            return 1

        i = 0
        while i < nlen:
            num = nums[i]

            if i + 1 == num:
                i += 1
                continue

            if num > 0 and num <= nlen and nums[num - 1] != num:
                nums[num - 1], nums[i] = nums[i], nums[num - 1]
            else:
                i += 1

        for i, num in enumerate(nums):
            if num != i + 1:
                return i + 1
        return nlen + 1
